wrc20.md

```diff
@@ -0,0 +1,314 @@
+WRC20
+=====
+
+```k
+requires "kwasm-lemmas.k"
+```
+
+Lemmas
+------
+
+```k
+module WRC20-LEMMAS [symbolic]
+    imports KWASM-LEMMAS
+```
+
+This conversion turns out to be helpful in this particular proof, but we don't want to apply it on all KWasm proofs.
+
+```k
+    rule X /Int 256 => X >>Int 8
+```
+
+TODO: The two `#get` theorems below theorems handle special cases in this proof, but we should be able to use some more general theorems to prove them.
+
+```k
+    rule (#get(BM, ADDR) +Int (X +Int Y)) modInt 256 => (#get(BM, ADDR) +Int ((X +Int Y) modInt 256)) modInt 256       [simplification]
+    rule (#get(BM, ADDR) +Int X)           >>Int 8   => X >>Int 8 requires X modInt 256 ==Int 0 andBool #isByteMap(BM) [simplification]
+```
+
+TODO: The following theorems should be generalized and proven, and moved to the set of general lemmas.
+Perhaps using `requires N ==Int 2 ^Int log2Int(N)`?
+
+```k
+    rule X *Int 256 >>Int N => (X >>Int (N -Int 8))   requires  N >=Int 8 [simplification]
+
+    rule (#get(BM, IDX) +Int (X <<Int 8)) >>Int N => X >>Int (N -Int 8) requires #isByteMap(BM) andBool N >=Int 8 [simplification]
+
+    rule (Y +Int X *Int 256) >>Int N => (Y >>Int N) +Int (X >>Int (N -Int 8))   requires  N >=Int 8 [simplification]
+
+    rule (X <<Int 8) +Int (Y <<Int 16) => (X +Int (Y <<Int 8)) <<Int 8 [simplification]
+
+    rule (X <<Int N) +Int (Y <<Int M) => (X +Int (Y <<Int (M -Int N))) <<Int N
+      requires N <=Int M
+      [simplification]
+
+    rule (X <<Int N) +Int (Y <<Int M) => ((X <<Int (N -Int M)) +Int Y) <<Int M
+      requires N >Int M
+      [simplification]
+
+    rule (X +Int (Y <<Int M)) modInt N => X +Int ((Y <<Int M) modInt N)
+      requires N ==Int (1 <<Int log2Int(N))
+       andBool N >=Int 256
+       andBool 0 <=Int X
+       andBool X <Int 256
+       andBool M >=Int 8
+      [simplification]
+```
+
+TODO: The following theorem should be proven, and moved to the set of general lemmas.
+
+```k
+    rule #getRange(BM, ADDR, WIDTH)  >>Int N => #getRange(BM, ADDR +Int 1, WIDTH -Int 1)  >>Int (N -Int 8)
+      requires N >=Int 8
+       andBool WIDTH >Int 1
+      [simplification]
+```
+
+```k
+endmodule
+```
+
+Macros
+------
+
+The following module gives macros for the wrc20 code so that its parts can be expressed succinctly in proofs.
+
+```k
+module WRC20
```

Comment:
The definitoin of the WRC20 is in this module. That used to be imported to KWASM-LEMMAS, but isn't anymore (good choice).
So either import both WRC20 and WRC20-LEMMAS in the WRC20-SPEC, or merge the WRC20 and the WRC20-LEMMAS modules.